_posts/2019-04-23-autoregressive-flows/experimenting-with-autoregressive-flows.Rmd

```diff
@@ -0,0 +1,645 @@
+- the Adam optimizer was used, with a learning rate of 1e-4;
+- there were two hidden layers for each MAF, with 100 units each;
+- training went on until no improvement occurred for 30 consecutive epochs on the validation set; and 
+- the base distribution was a multivariate Gaussian.
+
+This is all useful information for our attempt to estimate this dataset, but the essential bit is this. In case you knew the dataset already, you might have been wondering how the authors would deal with the dimensionality of the data: It is a time series, and the MADE architecture explored above introduces autoregressivity between features, not time steps. So how is the additional temporal autoregressivity to be handled? The answer is: The time dimension is essentially removed. In the authors' words, 
+
+> [...] it is a time series but was treated as if each example were an i.i.d. sample from the marginal distribution. 
+
+This undoubtedly is useful information for our present modeling attempt, but it also tells us something else: We might have to look beyond MADE layers for actual time series modeling.
+
+Now though let's look at this example of using MAF for multivariate modeling, with no time or spatial dimension to be taken into account.
+
+Following the hints the authors gave us, this is what we do.
+
+```{r}
+# load libraries -------------------------------------------------------------
+library(tensorflow)
+library(tfprobability)
+
+tfe_enable_eager_execution(device_policy = "silent")
+
+library(tfdatasets)
+library(dplyr)
+library(readr)
+library(purrr)
+library(caret)
+
+# read data ------------------------------------------------------------------
+df <- read_table2("ethylene_CO.txt",
+                  skip = 1,
+                  col_names = FALSE)
+
+# we don't know if we'll end up with the same columns as the authors did,
+# but we try (at least we do end up with 8 columns)
+df <- df[,-(1:3)]
+hc <- findCorrelation(cor(df), cutoff = 0.985)
+df2 <- df[,-c(hc)]
+
+# scale
+df2 <- scale(df2)
+
+# train-test split
+n_rows <- nrow(df2) # 4208261
+train_ids <- sample(1:n_rows, 0.5 * n_rows)
+x_train <- df2[train_ids, ]
+x_test <- df2[-train_ids, ]
+
+# create datasets
+batch_size <- 100
+train_dataset <- tf$cast(x_train, tf$float32) %>%
+  tensor_slices_dataset %>%
+  dataset_batch(batch_size)
+
+test_dataset <- tf$cast(x_test, tf$float32) %>%
+  tensor_slices_dataset %>%
+  dataset_batch(nrow(x_test))
+```
+
+To construct the flow, the first thing needed is the base distribution.
+
+```{r}
+base_dist <- tfd_multivariate_normal_diag(loc = rep(0, ncol(df2)))
+```
+
+Now for the flow, by default constructed with batchnorm and permutation of feature order. 
+
+```{r}
+num_hidden <- 100
+dim <- ncol(df2)
+
+use_batchnorm <- TRUE
+use_permute <- TRUE
+num_mafs <-10
+num_layers <- 3 * num_mafs
+
+bijectors <- vector(mode = "list", length = num_layers)
```

Comment:
It might be worth a comment here to say that this list will contain the transforms that we'll compose together.

Reply:
check